### ISSUE TITLE
Could not obtained shared lock on image

### ISSUE THREAD

**Issue by kcgthb** [apptainer/singularity#174]
Hi Greg!

Just updated our installation to 2.1 and got an issue when executing a container stored on a NFS filesystem:

```
$ singularity --version
2.1
$ singularity shell ~/container.img
ERROR  : Could not obtained shared lock on image
$ grep /home/kilian /proc/mounts
10.210.19.102:/ifs/home/kilian /home/kilian nfs rw,relatime,vers=3,rsize=131072,wsize=524288,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,mountaddr=10.210.19.102,mountvers=3,mountport=300,mountproto=udp,local_lock=none,addr=10.210.19.102 0 0
```

If I move the container to our Lustre filesystem, locking works fine and I can use the container normally.

From a quick look at the code, it seems you're using `flock()` in [`sexec.c`](https://github.com/gmkurtzer/singularity/blob/9bfe3145a6b6cda139a4bfe2d63897ef9911d88c/src/sexec.c#L296), which is not supported on NFS mounts (details at https://stackoverflow.com/questions/22409780/flock-vs-lockf-on-linux)
Maybe it could be worth considering using `lockf()` instead?

**Comment by kcgthb** [comment]
Just an update to add that mounting the NFS filesystem with `local_lock=flock` fixes the problem. But it's not the default, and local locking won't prevent several NFS clients to use a container at the same time.

**Comment by gmkurtzer** [comment]
Heya Killian!

Oh dang! Well, in addition to messing up my tenses it seems my test system didn't have that issue. Bummer!! I don't believe that lockf() supports exclusive Vs. shared locks, both of which I am making use of here because prevent someone entering the container as read/write while many process can be opened in shared.

I think the best bet is to make this configurable:

allow image writable = yes/no

And if no, never obtain locks on the image as they would not be necessary.

Thanks!

**Comment by gmkurtzer** [comment]
Hi Killian,

I think this is fixed now in 2.1.2 and the upcoming 2.2 release. I'm gonna close this ticket but if it needs to be reopened, comment and let me know and I'll reopen it.

Thanks!
